server/src/sim.rs

```diff
@@ -127,6 +166,30 @@ impl Sim {
         }
     }
 
+    fn snapshot_voxel_node(&self, node: NodeId) -> save::VoxelNode {
+        let mut chunks = vec![];
+        let node_data = self.graph.get(node).as_ref().unwrap();
+        for vertex in Vertex::iter() {
+            if !self.modified_chunks.contains(&ChunkId::new(node, vertex)) {
+                continue;
+            }
+            let mut serialized_voxels = Vec::new();
+            let Chunk::Populated { ref voxels, .. } = node_data.chunks[vertex] else {
+                panic!("Unknown chunk listed as modified");
+            };
+            postcard_helpers::serialize(
+                &voxels.to_serializable(self.cfg.chunk_size),
+                &mut serialized_voxels,
+            )
+            .unwrap();
```

Comment:
This encoding is incorrect. The .proto says:

```
    // Dense 3D array of 16-bit material tags for all voxels in this chunk
    bytes voxels = 2;
```

The postcard output for "struct of `Vec` of `Material`" is not consistent with that. We should construct a `Vec<u8>` for use in `Chunk` directly from `&[Material]` by something like `materials.flat_map(|x| x.to_le_bytes()).collect()`, which happens to be very concise anyway.

Reply:
Thanks! I forgot that the intended output was already documented.

Reply:
One issue I'm noticing is that `x` is of type `Material`, not `u16`, so the manual serialization also requires an `as u16`. This is fine for serialization, but for deserialization, we need to turn the `u16` back into a `Material` enum, which Rust doesn't support on its own. I can think of a few options:
- Import a new library to do this.
- Remove the `Material` enum entirely in favor of something more flexible (store a `u16` ourselves instead of using `repr`)
- Create a huge `match` statement to do the conversion (which I'm currently doing in my draft code)
- Use `unsafe` to do the conversion
- Use serde somehow to do the conversion.

I'm currently in favor of the second option, although I don't know whether to block this PR behind such a refactor.

Reply:
A big `match`, probably in a `try_from`, is the idiomatic way to do this. LLVM is known to optimize those out reliably. You can use a simple macro to generate them if it gets unwieldly.

Reply:
A crate like `int-enum` or `enumn` could prove useful here. I'll shy away from using a macro for now, but I'll use a unit test to enforce consistency on what would otherwise be a DRY-violation.

Reply:
`enumn` looks pleasant to me, but I don't mind handwritten either. DRY can be taken too far.

> I expect we'll want to switch away from an enum for other reasons (for instance, a dynamically-generated materials list with procedural textures and/or mod scripting support)

Oh, that's actually a really good point. Using a (newtyped) u16 would let us make materials totally data-driven, which is compelling for ease of maintenance and tinkering if nothing else. No need to pursue that in this PR though.

Reply:
> DRY can be taken too far.

This is definitely something I've had a change of heart on, as I used to consider DRY to be much more important. I think repeated logic is entirely fine as long as there's something enforcing consistency between the two versions, as avoiding inconsistencies between the two (that would cause complexity if intentional and bugs if not) is one of the biggest reasons for DRY as far as I understand.

Reply:
I decided to fix this by having the resulting `Vec<u8>` be present in `SerializableVoxelData`, which I consequently renamed to `SerializedVoxelData`. The main reason for this is that the only purpose of `SerializableVoxelData` is to be able to serialize it, and having different parts of the serialization in different parts of the code seems silly if both parts are done manually.

I decided to put this change in a separate commit, so this PR now has two commits.